### DISCUSSION TITLE
Question about middleware

### DISCUSSION THREAD

**Discussion by amsam0** [vercel/next.js#30903] [Help] [answered]
Am I correct that middleware runs on the client? I am unable to access server-side stuff like `fs` and I'm wondering if there is a way to make it run on the server and have similar execution behavior to `getServerSideProps`.

**Comment by shaneosullivan** [comment, marked as answer]
Middleware runs on the server but with a more limited set of APIs than you get on your normal server code.  These more limited APIs allow it to run on the Vercel Edge network which is fast and closer to the client.  So, it's about tradeoffs.

**Reply by amsam0** [reply]
Thanks for the help!

**Reply by dbismut** [reply]
@shaneosullivan I'm trying to import a package inside the middleware, and it looks like it's importing the `package.json` `browser` file for that package (which results in an error). Is that expected?

**Reply by shaneosullivan** [reply]
I can't say - I haven't tried that one (just a user of NextJS here, not on the team).

**Reply by dbismut** [reply]
No worries! Thanks anyways
